declarations/index.d.ts

```diff
@@ -1,22 +1,107 @@
+    delay?: number;
+}
+export interface ISerialConfig {
+    portId: number | string;
+    baud?: number;
+    rxPin?: number | string;
+    txPin?: number | string;
+}
+export interface IServoConfig {
+    pin: number | string;
+    min?: number;
+    max?: number;
 }
-export declare class AbstractIO extends EventEmitter implements IAbstractIO {
-    isReady: boolean;
+export declare type ReadHandler = (data: number[]) => void;
+export declare class AbstractIO extends EventEmitter {
     readonly MODES: {
-        INPUT: number;
-        OUTPUT: number;
-        ANALOG: number;
-        PWM: number;
-        SERVO: number;
+        INPUT: Mode;
+        OUTPUT: Mode;
+        ANALOG: Mode;
+        PWM: Mode;
+        SERVO: Mode;
+        UNKOWN: Mode;
     };
+    readonly HIGH: Value;
+    readonly LOW: Value;
+    readonly pins: IPinConfiguration[];
+    readonly analogPins: number[];
+    readonly name: string;
+    readonly defaultLed: string | number | undefined;
+    readonly isReady: boolean;
+    pinMode(pin: string | number, mode: Mode): void;
+    pwmWrite(pin: string | number, value: number): void;
+    servoWrite(pin: string | number, value: number): void;
+    digitalWrite(pin: string | number, value: number): void;
+    i2cWrite(address: number, inBytes: number[]): void;
+    i2cWrite(address: number, register: number, inBytes: number[]): void;
+    i2cWriteReg(address: number, register: number, value: number): void;
+    serialWrite(portId: string | number, inBytes: number[]): void;
+    analogRead(pin: string | number, handler: (value: number) => void): void;
+    digitalRead(pin: string | number, handler: (value: Value) => void): void;
+    i2cRead(address: number, bytesToRead: number, handler: ReadHandler): void;
+    i2cRead(address: number, register: number, bytesToRead: number, handler: ReadHandler): void;
+    i2cReadOnce(address: number, bytesToRead: number, handler: ReadHandler): void;
+    i2cReadOnce(address: number, register: number, bytesToRead: number, handler: ReadHandler): void;
+    pingRead(settings: IPingReadSettings, handler: (duration: number) => void): void;
+    serialRead(portId: number | string, handler: ReadHandler): void;
+    serialRead(portId: number | string, maxBytesToRead: number, handler: ReadHandler): void;
+    i2cConfig(options: II2CConfig): void;
+    serialConfig(options: ISerialConfig): void;
+    servoConfig(options: IServoConfig): void;
+    servoConfig(pin: number, min: number, max: number): void;
+    serialStop(portId: number | string): void;
+    serialClose(portId: number | string): void;
+    serialFlush(portId: number | string): void;
+    sendOneWireConfig(): void;
+    sendOneWireSearch(): void;
+    sendOneWireAlarmsSearch(): void;
+    sendOneWireRead(): void;
+    sendOneWireReset(): void;
+    sendOneWireWrite(): void;
+    sendOneWireDelay(): void;
+    sendOneWireWriteAndRead(): void;
+    setSamplingInterval(): void;
+    stepperConfig(): void;
+    stepperStep(): void;
```

Comment:
```stepperConfig()``` and ```stepperStep()``` depend on a custom build of configurable firmata, and there is a whole bunch of new stepper commands that fall into the same category.  We should either leave these two out of abstract-io or add in all the others. Pretty sure this holds true for pingRead too.

Reply:
Do we have documentation on what these methods are? Also, do you know if J5 does a test to see if the method exists? I brought these over from Raspi IO, which implements these two methods and throws an exception, but perhaps this isn't the correct approach?

Reply:
I was digging through the [J5 stepper code](https://github.com/rwaldron/johnny-five/blob/master/lib/stepper.js) a bit, and that code only seems to be calling `stepperConfig()` and `stepperStep()`. Am I missing other parts of the code that call into the IO plugin stepper code?

I did discover that `io.MODES` needs to have a `STEPPER` entry though. I'm going to set it to `5` in abstract-io. Does that sound good?

Reply:
The new stepper code for J5 is coming soon https://github.com/dtex/johnny-five/tree/stepper2  I had to get it into firmata protocol, configurableFirmata and firmata.js first.

Reply:
Thanks. To confirm, the complete list of new methods is:

`accelStepperConfig `
`accelStepperZero`
`accelStepperStep`
`accelStepperTo`
`accelStepperEnable`
`accelStepperStop`
`accelStepperReportPosition`
`accelStepperAcceleration`
`accelStepperSpeed`
`multiStepperConfig`
`multiStepperTo`
`multiStepperStop`

Did I miss any of them?

Reply:
No, that looks like all of them.